larq_compute_engine/core/bgemm_kernels_arm64.h

```diff
@@ -358,6 +362,26 @@ void BinaryKernelNeonOutOfOrder4x4(const BinaryKernelParams<4, 4>& params) {
       "fadd v26.4s, v26.4s, v15.4s\n"
       "fadd v27.4s, v27.4s, v15.4s\n"
 
+IF_INT8_OUTPUT(
+      // Convert to int32, rounding to nearest with ties to even.
+      "fcvtns v24.4s, v24.4s\n"
+      "fcvtns v25.4s, v25.4s\n"
+      "fcvtns v26.4s, v26.4s\n"
+      "fcvtns v27.4s, v27.4s\n"
+
+      // Narrow to int16 with "signed saturating extract narrow" instructions.
+      "sqxtn v24.4h, v24.4s\n"
+      "sqxtn v25.4h, v25.4s\n"
+      "sqxtn v26.4h, v26.4s\n"
+      "sqxtn v27.4h, v27.4s\n"
```

Comment:
Is it correct that these (together with the other 4 below) are only meant for clamping to int8 range? Or are they moving the values into the lower lanes?

Reply:
These are essentially equivalent to a saturating shift-right-and-narrow with a shift of 0 - in each register we start off with 4 int32s next to each other, and after this block we have 4 int16s next to each other in the lower half of the register. Then after the block below we end up with four bytes next to each other in the lower quarter of each register.

It's true that we could replace the block above with:
```cpp
      "sqxtn v24.4h, v24.4s\n"
      "sqxtn2 v24.8h, v25.4s\n"
      "sqxtn v25.4h, v26.4s\n"
      "sqxtn2 v25.8h, v27.4s\n"
```
and then the block below could become
```cpp
      "sqxtn v24.8b, v24.8h\n"
      "sqxtn v25.8b, v25.8h\n"
```
getting rid of two instructions. Then the values of interest are the first 8 bytes (lower half) of each of the two registers. But I don't think it's worth it because by using fewer registers we would have tighter dependency between nearby instructions which could cause some pipeline stalls (ideally we want > 3 cycles between writing to and reading from the same register).

Reply:
Ah and just to add, the clamping to int8 range is something that comes for free from using 'saturating' instructions - but we would still have to have these instructions even if not so as to gather the bytes correctly from the int32 values, unless we wanted to write one byte at a time to the destination.